CppCoreGuidelines.md

```diff
@@ -13926,29 +13926,52 @@ However, over time, code fragments can turn up in unexpected places.
 
     double cached_computation(double x)
     {
-        // bad: these two statics cause data races in multi-threaded usage
+        // bad: these statics cause data races in multi-threaded usage
         static double cached_x = 0.0;
         static double cached_result = COMPUTATION_OF_ZERO;
-        double result;
 
-        if (cached_x == x)
-            return cached_result;
-        result = computation(x);
-        cached_x = x;
-        cached_result = result;
-        return result;
+        if (cached_x != x) {
+            cached_x = x;
+            cached_result = computation(x);
+        }
+        return cached_result;
     }
 
 Although `cached_computation` works perfectly in a single-threaded environment, in a multi-threaded environment the two `static` variables result in data races and thus undefined behavior.
 
-There are several ways that this example could be made safe for a multi-threaded environment:
+##### Example, good
+
+    struct ComputationCache {
+        double cached_x = 0.0;
+        double cached_result = COMPUTATION_OF_ZERO;
+
+        double compute(double x) {
+            if (cached_x != x) {
+                cached_x = x;
+                cached_result = computation(x);
+            }
+            return cached_result;
+        }
+    };
+
+Here the cache is stored as member data of a `ComputationCache` object, rather than as shared static state.
+This refactoring essentially delegates the concern upward to the caller: a single-threaded program
+might still choose to have one global `ComputationCache`, while a multi-threaded program might
+have one `ComputationCache` instance per thread, or one per "context" for any definition of "context."
+The refactored function no longer attempts to manage the allocation of `cached_x`. In that sense,
+this is an application of the Single Responsibility Principle.
+
+In this specific example, refactoring for thread-safety also improved reusability in single-threaded
+programs. It's not hard to imagine that a single-threaded program might want two `ComputationCache` instances
+for use in different parts of the program, without having them overwrite each other's cached data.
```

Comment:
Is this on topic? There's kind of a circular thing here where we want to avoid global data because it makes multithreading harder (among other reasons) and the solution here to avoid global data to solve the multithreading issue has all the other benefits of avoiding global data. I wonder if a link to I.2 would suffice. However that may require #1599 to be addressed.

Reply:
I agree that #1599 is relevant. Re "on-topic," that depends on how much we want each guideline to stand isolated, versus how much this list is just presenting "different ways of looking at the elephant." You can approach the `ComputationCache` example from the thread-safety side, or the reusability side, or the Single-Responsibility-Principle side, but pretty much what it boils down to is "The same refactoring solves your problem no matter what you thought your problem was." :)

Also possible cross-links: C.1 and F.8 (if F.8 is generalized to talk about the hidden `*this` parameter, so that `cached_computation` is non-pure but `ComputationCache::compute` counts as pure).